pkg/sql/logictest/tests/local-leased-descriptors/local-leased-descriptors_test: TestLogic_schema_repair failed

pkg/sql/logictest/tests/local-leased-descriptors/local-leased-descriptors_test.TestLogic_schema_repair [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20845550?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20845550?buildTab=artifacts#/) on master @ [70a29e7123cdd3be72d47b391554f45ed581d765](https://github.com/cockroachdb/cockroach/commits/70a29e7123cdd3be72d47b391554f45ed581d765):


```
=== RUN   TestLogic_schema_repair
I251210 23:37:18.290531 1 (gostd) rand.go:288  [-] 1  random seed: -1501768686075673916
    test_log_scope.go:171: test logs captured to: /artifacts/tmp/_tmp/bb40179044e54071e884869168231d06/logTestLogic_schema_repair1916832679
    test_log_scope.go:82: use -show-logs to present logs inline
    logic.go:1606: apply knob opt logictest.knobOptAllowUnsafe
[23:37:18] setting distsql_workmem='18453B';
[23:37:19] --- done: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/17325/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/logictest/tests/local-leased-descriptors/local-leased-descriptors_test_/local-leased-descriptors_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/schema_repair with config local-leased-descriptors: 14 tests, 1 failures
--- total: 14 tests, 1 failures
    logic.go:4770: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/bb40179044e54071e884869168231d06/logTestLogic_schema_repair1916832679
--- FAIL: TestLogic_schema_repair (1.64s)
=== RUN   TestLogic_schema_repair/lost_table_data
    logic.go:3267: let $t_id = 107
    logic.go:3267: let $corrupt_id = 106
    logic.go:3267: let $parentID = 104
    logic.go:3267: let $parentSchemaID = 105
    logic.go:3267: let $json_t = {"table": {"columns": [{"id": 1, "name": "v", "nullable": true, "type": {"family": "IntFamily", "oid": 20, "width": 64}}, {"defaultExpr": "unique_rowid()", "hidden": true, "id": 2, "name": "rowid", "type": {"family": "IntFamily", "oid": 20, "width": 64}}], "createAsOfTime": {}, "families": [{"columnIds": [2], "columnNames": ["rowid"], "name": "fam_0_rowid"}, {"columnIds": [1], "columnNames": ["v"], "defaultColumnId": 1, "id": 1, "name": "fam_1_v"}], "formatVersion": 3, "id": 107, "modificationTime": {}, "name": "lostdata", "nextColumnId": 3, "nextConstraintId": 2, "nextFamilyId": 2, "nextIndexId": 2, "nextMutationId": 1, "parentId": 104, "primaryIndex": {"constraintId": 1, "createdAtNanos": "1765409839081664000", "encodingType": 1, "foreignKey": {}, "geoConfig": {}, "id": 1, "interleave": {}, "keyColumnDirections": ["ASC"], "keyColumnIds": [2], "keyColumnNames": ["rowid"], "name": "lostdata_pkey", "partitioning": {}, "sharded": {}, "storeColumnIds": [1], "storeColumnNames": ["v"], "unique": true, "vecConfig": {}, "version": 4}, "privileges": {"ownerProto": "root", "users": [{"privileges": "2", "userProto": "admin", "withGrantOption": "2"}, {"privileges": "2", "userProto": "root", "withGrantOption": "2"}], "version": 3}, "replacementOf": {"time": {}}, "schemaLocked": true, "unexposedParentSchemaId": 105, "version": "1"}}
    logic.go:3267: let $json_corrupt = {"table": {"columns": [{"id": 1, "name": "v", "nullable": true, "type": {"family": "IntFamily", "oid": 20, "width": 64}}, {"defaultExpr": "unique_rowid()", "hidden": true, "id": 2, "name": "rowid", "type": {"family": "IntFamily", "oid": 20, "width": 64}}], "createAsOfTime": {}, "families": [{"columnIds": [2, 1], "columnNames": ["rowid", "v"], "defaultColumnId": 1, "name": "fam_0_rowid_v"}], "formatVersion": 3, "id": 106, "modificationTime": {}, "name": "corruptdesc", "nextColumnId": 3, "nextConstraintId": 2, "nextFamilyId": 1, "nextIndexId": 2, "nextMutationId": 1, "parentId": 104, "primaryIndex": {"constraintId": 1, "createdAtNanos": "1765409839071894000", "encodingType": 1, "foreignKey": {}, "geoConfig": {}, "id": 1, "interleave": {}, "keyColumnDirections": ["ASC"], "keyColumnIds": [2], "keyColumnNames": ["rowid"], "name": "corruptdesc_pkey", "partitioning": {}, "sharded": {}, "storeColumnIds": [1], "storeColumnNames": ["v"], "unique": true, "vecConfig": {}, "version": 4}, "privileges": {"ownerProto": "root", "users": [{"privileges": "2", "userProto": "admin", "withGrantOption": "2"}, {"privileges": "2", "userProto": "root", "withGrantOption": "2"}], "version": 3}, "replacementOf": {"time": {}}, "schemaLocked": true, "unexposedParentSchemaId": 105, "version": "1"}}
    logic.go:3267: let $json_t_corrupt = {"table": {"columns": [{"id": 1, "name": "v", "nullable": true, "type": {"family": "IntFamily", "oid": 20, "width": 64}}, {"defaultExpr": "unique_rowid()", "hidden": true, "id": 2, "name": "rowid", "type": {"family": "IntFamily", "oid": 20, "width": 64}}], "createAsOfTime": {}, "families": [{"columnIds": [2, 1], "columnNames": ["rowid", "v"], "defaultColumnId": 1, "name": "fam_0_rowid_v"}], "formatVersion": 3, "id": 106, "modificationTime": {}, "nextColumnId": 3, "nextConstraintId": 2, "nextFamilyId": 1, "nextIndexId": 2, "nextMutationId": 1, "parentId": 104, "primaryIndex": {"constraintId": 1, "createdAtNanos": "1765409839071894000", "encodingType": 1, "foreignKey": {}, "geoConfig": {}, "id": 1, "interleave": {}, "keyColumnDirections": ["ASC"], "keyColumnIds": [2], "keyColumnNames": ["rowid"], "name": "corruptdesc_pkey", "partitioning": {}, "sharded": {}, "storeColumnIds": [1], "storeColumnNames": ["v"], "unique": true, "vecConfig": {}, "version": 4}, "privileges": {"ownerProto": "root", "users": [{"privileges": "2", "userProto": "admin", "withGrantOption": "2"}, {"privileges": "2", "userProto": "root", "withGrantOption": "2"}], "version": 3}, "replacementOf": {"time": {}}, "schemaLocked": true, "unexposedParentSchemaId": 105, "version": "1"}}
    logic.go:2398: 
         
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/17325/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/logictest/tests/local-leased-descriptors/local-leased-descriptors_test_/local-leased-descriptors_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/schema_repair:124: 
        expected success, but found
        (42601) relation "" (106): empty relation name
        errors.go:22: in ValidateName()
--- FAIL: TestLogic_schema_repair/lost_table_data (0.26s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLogic_schema_repair.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-57731

--- a/pkg/sql/catalog/descs/collection.go
+++ b/pkg/sql/catalog/descs/collection.go
@@ -39,6 +39,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/cockroach/pkg/util/iterutil"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/retry"
 	"github.com/cockroachdb/errors"
 )
 
@@ -123,6 +124,10 @@ type Collection struct {
 	// repairs.
 	skipValidationOnWrite bool
 
+	// waitForLockedLeaseBump indicates that we need to wait for the locked
+	// lease timestamp to bump as well.
+	waitForLockedLeaseBump bool
+
 	// readerCatalogSetup indicates that replicated descriptors can be modified
 	// by this collection.
 	readerCatalogSetup bool
@@ -189,6 +194,18 @@ func (tc *Collection) SkipValidationOnWrite() {
 	tc.skipValidationOnWrite = true
 }
 
+// MaybeSetLockedLeaseBump requires that the locked lease timestamp
+// is also bumped after this operation. This is needed in scenarios
+// where WaitForOneVersion has no prior version checked out, and the
+// previous version is invalid (i.e. has validation errors).
+func (tc *Collection) MaybeSetLockedLeaseBump(ctx context.Context) {
+	if !lease.LockedLeaseTimestamp.Get(&tc.settings.SV) ||
+		!tc.settings.Version.IsActive(ctx, clusterversion.V26_1) {
+		return
+	}
+	tc.waitForLockedLeaseBump = true
+}
+
 // SetReaderCatalogSetup indicates this collection is being used to
 // modify reader catalogs.
 func (tc *Collection) SetReaderCatalogSetup() {
@@ -199,6 +216,29 @@ func (tc *Collection) SetReaderCatalogSetup() {
 // the passed slice. Errors are logged but ignored.
 func (tc *Collection) ReleaseSpecifiedLeases(ctx context.Context, descs []lease.IDVersion) {
 	tc.leased.release(ctx, descs)
+}
+
+// MaybeWaitForLeaseTimestampBump waits for any lease timestamp bump, which
+// is normally used by repair queries to ensure the new version is available.
+// Normally for schema changes WaitForOneVersion is enough, since it guarantees
+// no one can lease the old version. However, if the prior version was invalid,
+// then there will be a delay after which the new version will be usable, as the
+// locked lease timestamp gets bumped.
+func (tc *Collection) MaybeWaitForLeaseTimestampBump(
+	ctx context.Context, commitTime hlc.Timestamp,
+) {
+	// This transaction does not require any leased timestamp bump.
+	if !tc.waitForLockedLeaseBump {
+		return
+	}
+	// Confirm the lease manager is leasing out any new descriptor versions
+	// at the commit time.
+	r := retry.StartWithCtx(ctx, retry.Options{})
+	for r.Next() {
+		if !tc.leased.lm.GetSafeReplicationTS().Less(commitTime) {
+			break
+		}
+	}
 }
 
 // ReleaseLeases releases all leases. Errors are logged but ignored.

--- a/pkg/sql/conn_executor.go
+++ b/pkg/sql/conn_executor.go
@@ -4228,6 +4228,11 @@ func (ex *connExecutor) txnStateTransitionsApplyWrapper(
 			if err := ex.waitForInitialVersionForNewDescriptors(cachedRegions); err != nil {
 				return advanceInfo{}, err
 			}
+			// If a repair query was executed, then we need to confirm that the lease manager
+			// is handing out the new descirptor version. This is covered by the previous waits
+			// if the prior version was valid, but if it was invalid then we need the lease manager
+			// to have a new timestamp available.
+			ex.extraTxnState.descCollection.MaybeWaitForLeaseTimestampBump(ex.Ctx(), advInfo.txnEvent.commitTimestamp)
 
 			execCfg := ex.planner.ExecCfg()
 			if err := UpdateDescriptorCount(ex.Ctx(), execCfg, execCfg.SchemaChangerMetrics); err != nil {

--- a/pkg/sql/repair.go
+++ b/pkg/sql/repair.go
@@ -170,6 +170,10 @@ func (p *planner) UnsafeUpsertDescriptor(
 	if force {
 		p.Descriptors().SkipValidationOnWrite()
 	}
+	// Indicate that for locked leasing we should wait for the
+	// locked lease timestamp to bump as well, in case the prior
+	// version of the descriptor was invalid.
+	p.Descriptors().MaybeSetLockedLeaseBump(ctx)
 
 	// If we are pushing out a brand new descriptor confirm that no leases
 	// exist before we publish it. This could happen if we did an unsafe delete,

